feat: generate endor compatible socket code

ereslibre opened the pull request:
A very primitive implementation. The example worker that was compiled with this modified code is at https://github.com/endorhq/endor/pull/86.

It has very basic socket primitives that use the endor networking support to reach internal services.

Angelmmiguel commented on src/library_sockfs.js:
```diff
@@ -5,6 +5,71 @@
+    addAtInit('}');
+  },
+  $ENDORFS__deps: [],
+  $ENDORFS: {
```
This is looking really nice! If we can have a fully separate library, that would be great. Two comments here:

* This change will require to add a new dependency to certain methods in the `library_syscall.js` file. I only found 2 occurrences of `SOCKFS` in the file, so I think it's totally fine. I just want to ensure we minimize the changes on that file.
* I would call this library `ENDOR_SOCKFS` to highlight this is an "alternative" version for SOCKFS.

Angelmmiguel commented on examples/tcp-client-example/Makefile:
```diff
@@ -1,6 +1,8 @@
+.PHONY: a.out.js
 a.out.js: main.c
-	../../emcc $< -o $@
+	../../emcc -g -sASYNCIFY -sSOCKET_DEBUG -sSYSCALL_DEBUG $< -o $@
```
For the final PR, I would move the debug flags to a separate make target, like `make debug`. Or make them optional with some environment variable.

Angelmmiguel commented on examples/tcp-client-example/main.c:
```diff
@@ -61,16 +63,23 @@ int main(int argc,char *argv[])
+      bytes = read(sockfd, response + received, total - received);
+      if (errno == EAGAIN) {
+        printf("EAGAIN\n");
+        emscripten_sleep(1000);
```
This would be interesting when we test the PHP build. Not sure how PHP will handle this case and if the process will block our code. Exciting to see if PHP works out of the box! 😄

Angelmmiguel commented on src/library_sockfs.js:
```diff
@@ -5,6 +5,156 @@
+    },
+    createSocket(family, type, protocol) {
+      type &= ~{{{ cDefs.SOCK_CLOEXEC | cDefs.SOCK_NONBLOCK }}}; // Some applications may pass it; it makes no sense for a single process.
+      // Endor only supports TCP
+      if ((type != {{{ cDefs.SOCK_STREAM }}}) || (protocol && protocol != {{{ cDefs.IPPROTO_TCP }}})) {
+        throw new FS.ErrnoError({{{ cDefs.EPROTONOSUPPORT }}});
+      }
```
Endor supports both UDP and TCP packets. However, I think it's better to start just with TCP. I would change this comment and add a TODO to implement UDP.

Angelmmiguel commented on src/library_sockfs.js:
```diff
@@ -5,6 +5,156 @@
+      }
+      return 'endor_socket[' + (ENDOR_SOCKFS.nextname.current++) + ']';
+    },
+    websocket_sock_ops: {
+      createPeer(sock, addr, port) {},
+      getPeer(sock, addr, port) {},
+      addPeer(sock, peer) {},
+      removePeer(sock, peer) {},
+      handlePeerEvents(sock, peer) {},
+    },
```
I guess this is a leftover and we can remove it.

ereslibre replied:
Ouch, right!

Angelmmiguel commented on src/library_sockfs.js:
```diff
@@ -5,6 +5,156 @@
+      },
+      connect(sock, addr, port) {
+        // Our universe --for now-- is ["192.168.10.20/32"]
+        if (addr != "192.168.10.20") {
+          throw new FS.ErrnoError({{{ cDefs.EHOSTUNREACH }}});
+        }
```
+1 for limiting this for now. However, I would be more flexible with the limitation as we might add more VMs in the future:

```suggestion
        if (!addr.startsWith("192.168.10.")) {
          throw new FS.ErrnoError({{{ cDefs.EHOSTUNREACH }}});
        }
```

Angelmmiguel commented on src/library_sockfs.js:
```diff
@@ -5,6 +5,156 @@
+      },
+      listen(sock, addr, port) {
+        // Client-side only for now
+        throw new FS.ErrnoError({{{ cDefs.EOPNOTSUPP }}});
```
Cool! That's a really great approach to make these operations as not supported 😄

Angelmmiguel reviewed (approved):
This initial implementation LGTM! Amazing work 😄 

Just added some minor comments.

ereslibre commented:
@Angelmmiguel thanks! Addressed all your comments. -O3 on our current emscripten version was not working due to a problem with `wasm-opt` and the flags it is invoked with. Given we have an input for emscripten I just changed the source of that input to the PR where emscripten is bumped to a version that fixes the wasm-opt problem.

I also checked that the our example program works as expected with this new version.
